Add the parent directory to changes set if new files are generated

mountkin opened the pull request:
The "TestChangesWithChanges" case randomlly fails on my development VM with the following errors:

```
--- FAIL: TestChangesWithChanges (0.00s)
        changes_test.go:201: no change for expected change C /dir1/subfolder != A /dir1/subfolder/newFile
```

The test details can be found here: http://pastebin.com/bYupfSr4

If I apply the following patch to changes_test.go, the test passes.

``` diff
diff --git a/pkg/archive/changes_test.go b/pkg/archive/changes_test.go
index 290b2dd..ba1aca0 100644
--- a/pkg/archive/changes_test.go
+++ b/pkg/archive/changes_test.go
@@ -156,6 +156,7 @@ func TestChangesWithChanges(t *testing.T) {
        }
        defer os.RemoveAll(layer)
        createSampleDir(t, layer)
+       time.Sleep(5 * time.Millisecond)
        os.MkdirAll(path.Join(layer, "dir1/subfolder"), 0740)

        // Let's modify modtime for dir1 to be sure it's the same for the two layer (to not having false positive)
```

It seems that if a file is created immediately after the directory is created, the `archive.Changes` function could't recognize that the parent directory of the new file is modified.

Perhaps the problem may reproduce on machines with low time precision?
I had successfully reproduced the failure on my development VM as well as a VM on DigitalOcean.

Signed-off-by: Shijiang Wei mountkin@gmail.com

mountkin committed:
Add the parent directory to changes set if new files are generated

The "TestChangesWithChanges" case randomlly fails on my development
VM with the following errors:
```
--- FAIL: TestChangesWithChanges (0.00s)
        changes_test.go:201: no change for expected change C /dir1/subfolder != A /dir1/subfolder/newFile
```

If I apply the following patch to changes_test.go, the test passes.

```diff
diff --git a/pkg/archive/changes_test.go b/pkg/archive/changes_test.go
index 290b2dd..ba1aca0 100644
--- a/pkg/archive/changes_test.go
+++ b/pkg/archive/changes_test.go
@@ -156,6 +156,7 @@ func TestChangesWithChanges(t *testing.T) {
        }
        defer os.RemoveAll(layer)
        createSampleDir(t, layer)
+       time.Sleep(5 * time.Millisecond)
        os.MkdirAll(path.Join(layer, "dir1/subfolder"), 0740)

        // Let's modify modtime for dir1 to be sure it's the same for the two layer (to not having false positive)
```

It seems that if a file is created immediately after the directory is created,
the `archive.Changes` function could't recognize that the parent directory of
the new file is modified.

Perhaps the problem may reproduce on machines with low time precision?
I had successfully reproduced the failure on my development VM as well as
a VM on DigitalOcean.

Signed-off-by: Shijiang Wei <mountkin@gmail.com>

vdemeester commented:
Hi @mountkin, I did reproduce this when running it with something like goconvey (which runs a lots of tests at the same times). I think it would probably be better to just add the ``time.Sleep()`` in the test rather than updating the code ; but we'll wait for other opinions.

mountkin commented:
My god! While this PR fixed the test on my development VM, it broke the jenkins test. 
Perhaps the "/dir1/subfolder" is added to the changes set twice?

@vdemeester At first I also thought adding the sleep is ok. But I'm not sure if it's a potential bug of the `Changes` function.

mountkin commented:
:laughing: Tests passed both on my VM and with janky.
ping @vbatts

vbatts commented:
oh sneaky. let me look.

vbatts commented:
LGTM

vbatts commented:
@duglin care to review?

duglin commented:
@vbatts okie dokie will do

duglin commented:
I'm kind of confused by this change. During the walk() call, why isn't the new dir automatically added? Is it not seen during the walk? The code in the walkFunc seems like it would add new dirs with a change type of "Add".  Or, due to some weird issue, are we actually walking to a file w/o walking to its parent dir?

duglin commented on pkg/archive/changes.go:
```diff
@@ -123,8 +127,16 @@ func Changes(layers []string, rw string) ([]Change, error) {
+					changedDirs[path] = struct{}{}
 					change.Kind = ChangeModify
 					break
+				} else if os.IsNotExist(err) && f.Mode().IsRegular() {
```
Should this check be done outside of the loop over the layers? Meaning, should we only add this file's parent dir if we don't see it in ANY of the parent layers? Is it really safe to do it just based on the first layer we look at?

duglin replied:
Also, if I'm following all of this correctly, we're trying to fix the case where we walk to a file but not its parent dir (weird, but ok), right? If so, it seems like:
1 - we should check to see if all parents are in the change list, not just the first parent dir
2 - we should check irrespective of whether 'f' is a file or a dir - if there's a really a bug that makes it we can walk to a file w/o seeing its dir, then it seems like we could also walk to a dir w/o see its parent dir too.
I guess what I'm saying is.... should this code here be rewritten as:
```
tmpChanges := []Change{}
for parent := filepath.Dir( path ) ; parent != "/" ; parent = filepath.Dir(parent) {
  if changedDirs[parent] != nil {
    break
  }
  tmpChanges = append(tmpChanges, Change{Path: parent, Kind: ChangeModify}
  changedDirs[parent] = struct{}{}
}
for i := len(tmpChanges) ; i > 0 ; i-- {
  changes = append(changed, tmpChanges[i-1]
}
changedDirs[path] = struct{}{}
```
and moved to right before the `changes = append(changes, change)` line ?

But I may not be following what going on  :-)

mountkin replied:
Hi @duglin, yes we're trying to fix the case where `f.Size() == stat.Size() && f.Mode() == stat.Mode() && sameFsTime(f.ModTime(), stat.ModTime())` is true but there are indeed new files created in the RW layer. This might happen when the RW layer is created immediately after (or rather nearly the same time) that the RO layer is created (see the patch that I tried to apply to changes_test.go for details).

Before opening this PR I did the following experiments on ext4 filesystem.
```
root@local-dev:~# mkdir A
root@local-dev:~# stat -c '%y' A
2015-06-03 10:53:39.969511999 +0800
root@local-dev:~# mkdir A/B
root@local-dev:~# stat -c '%y' A
2015-06-03 10:53:48.009511999 +0800
root@local-dev:~# stat -c '%y' A/B
2015-06-03 10:53:48.009511999 +0800
root@local-dev:~# echo test >A/B/file.txt
root@local-dev:~# stat -c '%y' A/B
2015-06-03 10:54:07.193511999 +0800  # modify time changed
root@local-dev:~# stat -c '%y' A
2015-06-03 10:53:48.009511999 +0800  # unchanged
```
From the test results above we can see that when the new file A/B/file.txt is created, only the direct parent directory B is modified. The modify time of directory A is unchanged. So I think it's not necessary to recursively add all the parent directories to `changedDirs`.

mountkin replied:
The ` && f.Mode().IsRegular() ` should be removed as you had pointed out.

mountkin commented:
@duglin After some farther experiments I found that whenever files are created, modified or deleted, the modify time of the direct parent is always changed.
I'll updated the PR again as you had suggested, but only add the direct parent directory to the "changes" slice.
The test case "TestChangesWithChanges" should be modified accordingly.

duglin commented:
@mountkin I just want to make sure I fully understand the situation, sorry if these are silly questions...
- is all of this due to a bug in the filesystem code where the timestamp of the parent dir isn't modified right away (or perhaps at all)? Previously someone mentioned that adding a sleep to the test seemed to fix it, so that's why I'm wondering if this a timing issue/bug with the FS
- the walk() call is traversing the files in the R/W (top) layer, right? During that processing, do you know if it hits the parent dir of the changed file? I'm wondering if the parent dir is present in the R/W layer but just w/o an unmodified timestamp?
- if the parent dir is present in the R/W layer, is just the immediate parent there, or are all of the parents there up to / ?

mountkin commented:
@duglin the "someone" is definitely me, see the patch to changes_test.go in the description of this PR :smile: 
I'm also curious that why the time of the directory is not modified if I don't add the sleep in the test code.

For your second question, according the godocs http://godoc.org/path/filepath#Walk,  "files are walked in lexical order", so the parent directory should be hit first.

The third question: I think all the parents should be there.

duglin commented on pkg/archive/changes_test.go:
```diff
@@ -178,6 +178,7 @@ func TestChangesWithChanges(t *testing.T) {
 	sort.Sort(changesByPath(changes))
 
 	expectedChanges := []Change{
+		{"/dir1", ChangeModify},
```
Since I'm not able to reproduce the bug, and the testcase works for me w/o this test change, I'd like @vbatts to verify that we should actually see `/dir1` in the list of changes.

Wish I would reproduce this - would love to debug some more to better understand it.  But if @vbatts is ok with the test change then I think the changes look good.

duglin commented:
Thanks to @mountkin for giving me a VM to reproduce this, I was able to get more info:
```
t1: "2015-06-03 11:51:16.829413949 -0400 EDT"
t2: "2015-06-03 11:51:16.829980647 -0400 EDT"
```
those are the real time before we create the two subfolders on each layer
```
Time: "2015-06-03 11:51:16.826944999 -0400 EDT"
Time: "2015-06-03 11:51:16.826944999 -0400 EDT"
```
those are the stat times for the two subfolders on each layer.  Notice that stat() is returning a less precise/rounded time for each dir.  So, when we check to see if the dirs have been modified, if we didn't cross the threshold for the rounding, then it'll appear as though they didn't change.

I think the chances of this happening in a real environment are pretty slim - you would have to create two layers and two same-named dirs on both layers within that rounding threshold.  Pretty slim.  

However, I think the code changes being proposed are still ok to make since it should be a no-op.  The reason we'd need to change the testcase (to add 'dir1' to the list of changes) is because the testcase goes out of its way to set the timestamp on 'dir1' on both layers to be the same value. Not sure why.  @vbatts do you have any insight into that one?  

@vbatts if you're ok with the testcase having one more item in the list of changes then I'm ok too.  I think the only other thing we should do is modify the testcase to NOT set 'dir1's mtime since that'll just confuse people.

vdemeester commented:
@duglin well.. I wrote this test :sweat:. I did add this comment *"Let's modify modtime for dir1 to be sure it's the same for the two layer (to not having false positive)"*. But it turns out it's wrong :disappointed:, got false positive even..

If I remember correctly, this was to be sure it would detect the change inside ``dir1`` without detecting a ``dir1`` change. But the more I look at this, the more ugly it seems to me (the way I did it :sob:).

duglin commented:
@vdemeester I don't think 'dir1' is the issue. Granted, with this PR I think modifying its timestamp isn't necessary (and could be confusing) but the bigger issue is why stat()'s precision on the mtime is so bad? Probably not a lot we can do about it, but at least this PR seems to at least work around it.

mountkin commented:
I just removed the `os.Chtimes` block.
According my experiment results when a file is deleted from or added to a directory, the mtime of the directory is changed. Since /dir1/file1-2 is deleted, the mtime of dir1 should change, so it must be in the `changes` slice.

duglin commented:
If I have /dir1/dir2/dir3/file.txt in a base layer. Then in the next layer I touch file.txt. In a "normal" environment, "dir3" will have its mtime changed and this PR shouldn't be needed - which means hopefully it'll be a no-op, right?.  However, would this change make it so that 'dir2' is in the change list?

I just modified the testcase to try this out. Perhaps I did something wrong but I only see file.txt in the change list w/o this PR. With this PR I see dir3 as well.  I even added a sleep(3) before I touched file.txt to make sure we didn't hit a timing issue.

@vbatts should 'dir3' be in the change list irrespective of this PR?

duglin commented:
ok in my testing a dir's mtime is only modified when a new file appears, or disappears, under it, not when a file under it is changed.   I wonder if we need to take into account the type of change of the original file in question before we add its parent.

duglin commented:
So, back to my original question... i think this PR does in fact add 'dir2' by mistake.  By modifying the testcase so it adds a new file instead of touching a file (in dir3) I can make 'dir2' appear in the change list but I don't think it should be.

Not sure yet what the right fix is, so for now just FYI...

duglin commented:
ok just wrappering the new logic that adds the parent dir with
```
if change.Kind == ChangeAdd || change.Kind == ChangeDelete {
```
seems to show the correct results.

duglin commented:
clearly need more tests for this too :-)

mountkin commented:
@duglin When I added `if change.Kind == ChangeAdd || change.Kind == ChangeDelete {` and modified the test case, it passed outside of the test infrastructure but failed inside the test infrastructure.
The reason that janky failed is that `f.Size()` and `stat.Size()` are not equal when testing in docker.

"rwPath: /dir2, roPath: /tmp/docker-changes-test-layer911993879/dir2, **f.Size: 60, stat.Size: 80**, f.Mode: 2147484096, stat.Mode: 2147484096, f.ModTime: 2015-06-08T15:56:22.937557991Z, stat.ModTime: 2015-06-08T15:56:22.937557991Z"

After digging the godocs I found the following comment in the `os.FileInfo` struct (http://godoc.org/os#FileInfo).
```go
Size() int64        // length in bytes for regular files; system-dependent for others
```

So my conclusion: the parent directory should be added to the change list irrespective of the kind of the change.

vbatts commented:
ok. so where are we on this? review needed now or still refining?
Also, there look to be a couple of failures in the CI

duglin commented:
@mountkin I think the testcase needs to be changed.
See my branch here: https://github.com/duglin/docker/tree/Changes
and let me know if you think my thinking is wrong.

mountkin commented:
Thanks @duglin I think your test case is reasonable. I just cherry-picked your test case to this PR. 
But after squash it seems that your name isn't showing in the commit logs. Sorry for that.

duglin commented:
@mountkin no worries, as long as the correct code gets in, that's all that matters.

@vbatts @vdemeester you guys ok with the changes?  Please check the testcases too since we did modify those a little.  I want to make sure we don't incorrectly add, or remove, from the list of changes people are used to seeing.

duglin commented:
LGTM

vdemeester commented:
LGTM

vbatts commented:
LGTM
